### PR TITLE
Push driver config during `Init`

### PR DESCRIPTION
**Pull request by mrjana**
Currently the driver configuration is pushed through a separate
api. This makes driver configuration possible at any arbitrary
time. This unncessarily complicates the driver implementation.
More importantly the driver does not get access to it's
configuration before it can do the handshake with libnetwork.
This make the internal drivers a little bit different to
external plugins which can get their configuration before the handshake
with libnetwork.

This PR attempts to fix that mismatch between internal drivers and
external plugins.

Signed-off-by: Jana Radhakrishnan <mrjana@docker.com>

### CODE REVIEW

**Comment by tomdee** [comment]
Are there any changes to the remote API in this PR? If so, I don't see any updates to remote.md

**Comment by mrjana** [comment]
@tomdee There are no changes to remote API because of this.

File: drivers_linux.go [hunk 1]
```diff
@@ -1,25 +1,76 @@
...
+		{remote.Init, "remote"},
+		{o.Init, "overlay"},
+	}
+
+	for _, i := range initializers {
+		updateDriverConfig(c, i.ntype, config, true)
+		if err := i.fn(c, config); err != nil {
 			return err
 		}
+		updateDriverConfig(c, i.ntype, config, false)
```
Comment (mavenugo) [inline drivers_linux.go]: shall we remove this call ? and rather initialize `config := make(map[string]interface{})` inside the for.
that makes the code easily readable.

File: drivers_linux.go [hunk 2]
```diff
@@ -1,25 +1,76 @@
...
 	}
+
 	return nil
 }
+
+func updateDriverConfig(c *controller, ntype string, config map[string]interface{}, isAdd bool) {
```
Comment (mavenugo) [inline drivers_linux.go]: ofcourse, with the above comment, we dont need the `isAdd` flag and hence the code will get even simpler.

**Comment by mrjana** [comment]
@mavenugo Updated to take care of the comments. PTAL

File: drivers_linux.go [hunk 3]
```diff
@@ -1,25 +1,66 @@
...
+	}{
+		{bridge.Init, "bridge"},
+		{host.Init, "host"},
+		{null.Init, "null"},
+		{remote.Init, "remote"},
+		{o.Init, "overlay"},
+	}
+
+	for _, i := range initializers {
+		if err := i.fn(c, makeDriverConfig(c, i.ntype)); err != nil {
```
Comment (mavenugo) [inline drivers_linux.go]: Can you please make this change in drivers_windows.go and drivers_freebsd.go as well ?

File: drivers_linux.go [hunk 4]
```diff
@@ -1,25 +1,66 @@
...
+
+	config := make(map[string]interface{})
+
+	if c.validateDatastoreConfig() {
+		config[netlabel.KVProvider] = c.cfg.Datastore.Client.Provider
+		config[netlabel.KVProviderURL] = c.cfg.Datastore.Client.Address
+	}
+
+	for _, label := range c.cfg.Daemon.Labels {
+		if !strings.HasPrefix(label, netlabel.DriverPrefix+"."+ntype) {
```
Comment (mavenugo) [inline drivers_linux.go]: To be more precise, can you change `strings.HasPrefix(label` to `strings.HasPrefix(netlabel.Key(label)`...
Though the current code works by the virtue of how the netlabel is defined. But still for technical accuracy,
I think ading the netlabel.Key() makes it more easily readable. WDYT ?

**Comment by mrjana** [comment]
@mavenugo Updated again taking care of the comments. PTAL

**Comment by mavenugo** [comment]
LGTM

**Pull request merged**